File: Java/[Udemy] Java POO/session12/StringBUILDER_ex2/src/entities/Order.java
package entities;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import entities.enums.OrderStatus;

public class Order {
	
	private Client client;
	private LocalDateTime date;
	private OrderStatus status;
	
	private List<OrderItem> orders = new ArrayList<>();

	public Order(Client client, LocalDateTime date, OrderStatus status) {
		this.client = client;
		this.date = date;
		this.status = status;
	}

	public Client getClient() {
		return client;
	}

	public void setClient(Client client) {
		this.client = client;
	}

	public LocalDateTime getDate() {
		return date;
	}

	public void setDate(LocalDateTime date) {
		this.date = date;
	}

	public OrderStatus getStatus() {
		return status;
	}

	public void setStatus(OrderStatus status) {
		this.status = status;
	}

	public List<OrderItem> getOrders() {
		return orders;
	}
	
	public void addItem(OrderItem orderItem) {
        orders.add(orderItem);
    }
	
	public void removeItem(OrderItem orderItem) {
        orders.remove(orderItem);
    }
	
	public double total() {
		double total = 0.0;
        for(OrderItem item : orders) {
            total += item.subTotal();
        }
        return total;
	}
	
	@Override
    public String toString() {
		
		DateTimeFormatter fmt = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss");
		StringBuilder sb = new StringBuilder();
		
        sb.append(String.format("Order moment: %s%n", getDate().format(fmt)));
        sb.append(String.format("Order status: %s%n", getStatus()));
        sb.append(String.format("-----------------%n"));
        sb.append(String.format("Client: %s%n", getClient()));
        sb.append(String.format("-----------------%n"));
        sb.append(String.format("Order items:%n"));
        for(OrderItem item : orders) {
        	sb.append(String.format("%s%n", item));
        }
        sb.append(String.format("-----------------%n"));
        sb.append(String.format("Total price: $%.2f%n", total()));
        
        return sb.toString();
	}
}

File: CSharp/[Udemy] CSharp POO/session9/Session90Ex3/Entities/Order.cs
using Session90Ex3.Entities.Enums;
using System.Globalization;
using System.Text;

namespace Session90Ex3.Entities
{
    internal class Order
    {
        public DateTime Moment { get; set; }
        public OrderStatus Status { get; set; }
        public Client Client { get; set; }
        public List<OrderItem> Items { get; set; } = new List<OrderItem>();
        public Order()
        {
        }
        public Order(DateTime moment, OrderStatus status, Client client)
        {
            Moment = moment;
            Status = status;
            Client = client;
        }
        public void AddItem(OrderItem item)
        {
            Items.Add(item);
        }
        public void RemoveItem(OrderItem item)
        {
            Items.Remove(item);
        }
        public double Total()
        {
            double total = 0;
            foreach (OrderItem item in Items)
            {
                total += item.SubTotal();
            }

            return total;
        }
        public override string ToString()
        {
            StringBuilder text = new StringBuilder();
            text.AppendLine($"Order moment: {Moment.ToString()}");
            text.AppendLine($"Order status: {Status}");
            text.AppendLine($"Client: {Client.Name} ({Client.BirthDate.ToShortDateString()}) - {Client.Email}");
            text.AppendLine("Order items:");
            foreach (OrderItem item in Items)
            {
                double total = 0;
                text.AppendLine($"{item.Product.Name}, ${item.Product.Price.ToString("F2", CultureInfo.InvariantCulture)}, Quantity: {item.Quantity}, Subtotal: ${item.SubTotal().ToString("F2", CultureInfo.InvariantCulture)}");
                text.AppendLine($"Total price: ${Total().ToString("F2", CultureInfo.InvariantCulture)}");
            }

            return text.ToString();
        }
    }
}
